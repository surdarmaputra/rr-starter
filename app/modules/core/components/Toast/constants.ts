import {
  CircleAlert,
  CircleCheck,
  Info,
  type LucideProps,
  TriangleAlert,
} from 'lucide-react';
import type { ForwardRefExoticComponent, RefAttributes } from 'react';

export const VARIANT = {
  default: 'default',
  success: 'success',
  warning: 'warning',
  danger: 'danger',
} as const;

export type VariantType = (typeof VARIANT)[keyof typeof VARIANT];

export const BACKGROUND_CLASSNAME: Record<VariantType, string> = {
  [VARIANT.default]: 'bg-white dark:bg-accent ring-divider',
  [VARIANT.success]:
    'bg-toast-success-background ring-toast-success-border shadow-success-disabled',
  [VARIANT.warning]:
    'bg-toast-warning-background ring-toast-warning-border shadow-warning-disabled',
  [VARIANT.danger]:
    'bg-toast-danger-background ring-toast-danger-border shadow-danger-disabled',
};

export const TITLE_CLASSNAME: Record<VariantType, string> = {
  [VARIANT.default]: 'text-letter-title',
  [VARIANT.success]: 'text-toast-success-title',
  [VARIANT.warning]: 'text-toast-warning-title',
  [VARIANT.danger]: 'text-toast-danger-title',
};

export const ICON_CLASSNAME: Record<VariantType, string> = {
  [VARIANT.default]: 'fill-backdrop text-letter-inverted',
  [VARIANT.success]: 'fill-toast-success-title',
  [VARIANT.warning]: 'fill-toast-warning-title',
  [VARIANT.danger]: 'fill-toast-danger-title',
};

export const ICON: Record<
  VariantType,
  ForwardRefExoticComponent<
    Omit<LucideProps, 'ref'> & RefAttributes<SVGSVGElement>
  >
> = {
  [VARIANT.default]: Info,
  [VARIANT.success]: CircleCheck,
  [VARIANT.warning]: TriangleAlert,
  [VARIANT.danger]: CircleAlert,
};

export const DESCRIPTION_CLASSNAME: Record<VariantType, string> = {
  [VARIANT.default]: 'text-letter-caption',
  [VARIANT.success]: 'text-toast-success-title opacity-60',
  [VARIANT.warning]: 'text-toast-warning-title opacity-70',
  [VARIANT.danger]: 'text-toast-danger-title opacity-60',
};

export const CLOSE_BUTTON_CLASSNAME: Record<VariantType, string> = {
  [VARIANT.default]:
    'text-letter-disabled hover:bg-accent hover:text-letter-caption',
  [VARIANT.success]:
    'text-toast-success-title opacity-60 hover:bg-success-disabled hover:text-toast-success-title hover:opacity-100',
  [VARIANT.warning]:
    'text-toast-warning-title opacity-70 hover:bg-warning-disabled hover:text-toast-warning-title hover:opacity-100',
  [VARIANT.danger]:
    'text-toast-danger-title opacity-60 hover:bg-danger-disabled hover:text-toast-danger-title hover:opacity-100',
};
