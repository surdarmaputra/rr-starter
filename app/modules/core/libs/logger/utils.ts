import { isbot } from 'isbot';
import { isRouteErrorResponse } from 'react-router';

import {
  ERROR_LOG_MESSAGE,
  INFO_LOG_MESSAGE,
  WARNING_LOG_MESSAGE,
} from './constants';
import { logger } from './logger';

export function logIncomingRequest(request: Request, isSpaMode?: boolean) {
  const userAgent = request.headers.get('user-agent');
  const isBot = userAgent && isbot(userAgent);

  logger.logInfo(INFO_LOG_MESSAGE.INCOMING_REQUEST, {
    is_bot: isBot,
    is_spa_mode: isSpaMode,
    url: request.url,
    user_agent: userAgent,
  });
}

export function logRequestError(error: unknown) {
  if (isRouteErrorResponse(error) && error.status === 404) {
    logger.logWarn(WARNING_LOG_MESSAGE.ROUTE_NOT_FOUND, error);
  } else {
    logger.logError(ERROR_LOG_MESSAGE.GENERAL_SERVER_ERROR, error);
  }
}

export function wrapWithApiLogger(
  request: Request,
  handler: () => Promise<Response>,
) {
  try {
    logIncomingRequest(request);
    return handler();
  } catch (error) {
    logRequestError(error);
  }
}
