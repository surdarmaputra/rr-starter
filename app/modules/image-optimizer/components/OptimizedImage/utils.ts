import {
  SCREEN_DENSITY_FACTORS,
  SCREEN_DENSITY_REDUCTION_MULTIPLIER,
  SUPPORTED_IMAGE_FORMATS,
  type SupportedImageFormat,
  TAILWIND_BREAKPOINTS,
} from './constants';

export function hasOriginUrl(url: string): boolean {
  try {
    return new URL(url).origin !== 'null';
  } catch {
    return false;
  }
}

export function detectBestFormat(): SupportedImageFormat {
  const canvas = document.createElement('canvas');
  if (canvas.getContext && canvas.getContext('2d')) {
    for (const format of Object.values(SUPPORTED_IMAGE_FORMATS)) {
      if (
        canvas.toDataURL(`image/${format}`).indexOf(`data:image/${format}`) ===
        0
      ) {
        return format;
      }
    }
  }
  return 'jpeg';
}

export function generateOptimizedUrl(
  url: string,
  format: SupportedImageFormat,
  width: number,
): string {
  return `/api/optimize-image?url=${encodeURIComponent(url)}&width=${width}&format=${format}`;
}

export function generateSrcSet(
  url?: string,
  format?: SupportedImageFormat,
  width?: number,
): string {
  if (!url || !format) {
    return '';
  }

  const srcSetEntries: string[] = [];

  if (width) {
    // If width prop is provided, generate variations for different pixel densities
    for (const density of SCREEN_DENSITY_FACTORS) {
      const effectiveWidth = width * density;
      srcSetEntries.push(
        `${generateOptimizedUrl(url, format, effectiveWidth)} ${effectiveWidth}w`,
      );
    }
  } else {
    // If no width prop, generate variations for Tailwind breakpoints
    for (const breakpoint of TAILWIND_BREAKPOINTS) {
      for (const density of SCREEN_DENSITY_FACTORS) {
        const effectiveWidth = Math.round(
          breakpoint * density * SCREEN_DENSITY_REDUCTION_MULTIPLIER,
        );
        srcSetEntries.push(
          `${generateOptimizedUrl(url, format, effectiveWidth)} ${effectiveWidth}w`,
        );
      }
    }
  }

  return srcSetEntries.join(', ');
}

export function generateSizes(width?: number): string {
  if (width) {
    // If width prop is provided, return the fixed width in CSS pixels
    return `${width}px`;
  }

  // If no width prop, use Tailwind breakpoints for responsive sizing
  const sizesByBreakpoint = TAILWIND_BREAKPOINTS.map(
    (breakpoint) =>
      `(max-width: ${breakpoint}px) ${Math.round(breakpoint * SCREEN_DENSITY_REDUCTION_MULTIPLIER)}px`,
  );
  return [...sizesByBreakpoint, '100vw'].join(', ');
}
