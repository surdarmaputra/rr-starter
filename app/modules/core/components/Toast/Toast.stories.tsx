import type { Meta, StoryObj } from '@storybook/react';

import { Button } from '../Button/Button';
import { Toaster, type ToasterProps } from './Toast';
import { toast } from './utils';

type Story = StoryObj<typeof Toaster>;

const meta: Meta = {
  component: Toaster,
  title: 'Core/Components/Toast',
};

export default meta;

export const Default: Story = {
  parameters: {
    docs: {
      source: {
        code: `
        
import { toast } from 'sonner';
        
// JSX
<div className="flex h-64 flex-col gap-2">
  <div className="flex gap-2">
    <Button onClick={() => toast("This is default toast")}>
      Default Toast
    </Button>
    <Button
      onClick={() =>
        toast({
          title: "Closable Toast",
          isClosable: false,
        })
      }
    >
      Not Closable Toast
    </Button>
    <Button
      onClick={() =>
        toast(
          {
            description: "This is a top right toast",
            isClosable: true,
            title: "Top Right",
          },
          {
            position: "top-right",
          }
        )
      }
    >
      Toggle Top Right
    </Button>
    <Button
      onClick={() =>
        toast({
          description: "This is a toast with action",
          title: "Toast with action",
          action: (
            <Button size="sm" variant="primary">
              Action
            </Button>
          ),
        })
      }
    >
      With Action
    </Button>
    <Button
      onClick={() =>
        toast({
          description: "This is a toast with action",
          title: "Toast with action",
          action: ({ dismiss }) => (
            <Button onClick={dismiss} size="sm" variant="primary">
              Dismiss
            </Button>
          ),
        })
      }
    >
      With Action Renderer
    </Button>
  </div>

  <div className="flex gap-2">
    <Button
      onClick={() => toast.success("Congrats! Action success.")}
      variant="success"
    >
      Success Toast
    </Button>
    <Button
      onClick={() =>
        toast.success({
          description: "Congrats! Action success.",
          title: "Success Toast",
        })
      }
      variant="success"
    >
      Success Description
    </Button>
    <Button
      onClick={() => toast.warning("Watch out! This action might by harmful")}
      variant="warning"
    >
      Warning Toast
    </Button>{" "}
    <Button
      onClick={() =>
        toast.warning({
          description: "Watch out! This action might by harmful",
          title: "Warning Toast",
        })
      }
      variant="warning"
    >
      Warning Description
    </Button>
    <Button
      onClick={() => toast.danger("Oops! Something wrong.")}
      variant="danger"
    >
      Danger Toast
    </Button>{" "}
    <Button
      onClick={() =>
        toast.danger({
          description: "Oops! Something wrong.",
          title: "Error Toast",
        })
      }
      variant="danger"
    >
      Danger Description
    </Button>
  </div>
  <Toaster />
</div>
        `,
      },
    },
  },
  render: (args: ToasterProps) => (
    <div className="flex h-64 flex-col gap-2">
      <div className="flex gap-2">
        <Button onClick={() => toast('This is default toast')}>
          Default Toast
        </Button>
        <Button
          onClick={() =>
            toast({
              title: 'Closable Toast',
              isClosable: false,
            })
          }
        >
          Not Closable Toast
        </Button>
        <Button
          onClick={() =>
            toast(
              {
                description: 'This is a top right toast',
                isClosable: true,
                title: 'Top Right',
              },
              {
                position: 'top-right',
              },
            )
          }
        >
          Toggle Top Right
        </Button>
        <Button
          onClick={() =>
            toast({
              description: 'This is a toast with action',
              title: 'Toast with action',
              action: (
                <Button size="sm" variant="primary">
                  Action
                </Button>
              ),
            })
          }
        >
          With Action
        </Button>
        <Button
          onClick={() =>
            toast({
              description: 'This is a toast with action',
              title: 'Toast with action',
              action: ({ dismiss }) => (
                <Button onClick={dismiss} size="sm" variant="primary">
                  Dismiss
                </Button>
              ),
            })
          }
        >
          With Action Renderer
        </Button>
      </div>

      <div className="flex gap-2">
        <Button
          onClick={() => toast.success('Congrats! Action success.')}
          variant="success"
        >
          Success Toast
        </Button>
        <Button
          onClick={() =>
            toast.success({
              description: 'Congrats! Action success.',
              title: 'Success Toast',
            })
          }
          variant="success"
        >
          Success Description
        </Button>
        <Button
          onClick={() =>
            toast.warning('Watch out! This action might by harmful')
          }
          variant="warning"
        >
          Warning Toast
        </Button>{' '}
        <Button
          onClick={() =>
            toast.warning({
              description: 'Watch out! This action might by harmful',
              title: 'Warning Toast',
            })
          }
          variant="warning"
        >
          Warning Description
        </Button>
        <Button
          onClick={() => toast.danger('Oops! Something wrong.')}
          variant="danger"
        >
          Danger Toast
        </Button>{' '}
        <Button
          onClick={() =>
            toast.danger({
              description: 'Oops! Something wrong.',
              title: 'Error Toast',
            })
          }
          variant="danger"
        >
          Danger Description
        </Button>
      </div>
      <Toaster {...args} />
    </div>
  ),
};
