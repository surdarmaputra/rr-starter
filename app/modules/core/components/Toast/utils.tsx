import { type ExternalToast, toast as sonnerToast } from 'sonner';

import { Toast, type ToastProps } from './Toast';

function toastVariantRenderer(variant: ToastProps['variant']) {
  return (props: string | Omit<ToastProps, 'id'>, options?: ExternalToast) => {
    if (typeof props === 'string') {
      return sonnerToast.custom(
        (id) => <Toast id={id} title={props} variant={variant} />,
        options,
      );
    }
    return sonnerToast.custom(
      (id) => <Toast id={id} {...props} variant={variant} />,
      options,
    );
  };
}

export function toast(
  props: string | Omit<ToastProps, 'id'>,
  options?: ExternalToast,
) {
  return toastVariantRenderer('default')(props, options);
}

toast.success = toastVariantRenderer('success');
toast.warning = toastVariantRenderer('warning');
toast.danger = toastVariantRenderer('danger');
