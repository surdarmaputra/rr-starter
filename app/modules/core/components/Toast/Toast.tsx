import { X } from 'lucide-react';
import type { ReactNode } from 'react';
import { toast, Toaster } from 'sonner';

import { cn } from '~/libs/shadcn/utils';

import {
  BACKGROUND_CLASSNAME,
  CLOSE_BUTTON_CLASSNAME,
  DESCRIPTION_CLASSNAME,
  ICON,
  ICON_CLASSNAME,
  TITLE_CLASSNAME,
  VARIANT,
  type VariantType,
} from './constants';

interface ToastActionRendererParams {
  id?: string | number;
  dismiss?: () => void;
}

export type ToasterProps = React.ComponentProps<typeof Toaster>;

export interface ToastProps {
  id: string | number;
  title: ReactNode;
  description?: ReactNode;
  isClosable?: boolean;
  action?: ReactNode | ((params: ToastActionRendererParams) => ReactNode);
  variant?: VariantType;
}
function Toast({
  id,
  title,
  description,
  isClosable = true,
  action,
  variant = VARIANT.default,
}: ToastProps) {
  const IconComponent = ICON[variant];

  return (
    <div
      className={cn(
        'relative flex w-full items-center gap-2 rounded-lg p-4 shadow-lg ring-1 md:w-80',
        BACKGROUND_CLASSNAME[variant],
      )}
      role="alert"
    >
      <div className="h-6 w-6 shrink-0 self-start">
        <IconComponent
          className={cn('h-6 w-6 text-white', ICON_CLASSNAME[variant])}
        />
      </div>

      <div className="flex flex-1 items-center">
        <div className="w-full">
          <p
            className={cn(
              'pr-6 text-sm font-medium text-letter-body',
              TITLE_CLASSNAME[variant],
            )}
          >
            {title}
          </p>
          {Boolean(description) && (
            <p
              className={cn(
                'mt-1 pr-6 text-sm text-letter-caption',
                DESCRIPTION_CLASSNAME[variant],
              )}
            >
              {description}
            </p>
          )}
        </div>
      </div>

      {isClosable && !action && (
        <button
          className={cn(
            'absolute right-4 top-4 rounded-full bg-transparent p-1 text-letter-disabled hover:bg-accent hover:text-letter-caption',
            CLOSE_BUTTON_CLASSNAME[variant],
          )}
          onClick={() => toast.dismiss(id)}
        >
          <span className="sr-only">Close</span>
          <X aria-hidden="true" className="h-4 w-4" />
        </button>
      )}

      {typeof action === 'function'
        ? action({ id, dismiss: () => toast.dismiss(id) })
        : action}
    </div>
  );
}

export { Toast, Toaster };
