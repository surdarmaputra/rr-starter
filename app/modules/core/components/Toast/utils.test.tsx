import { render, screen, within } from '@testing-library/react';
import { beforeEach, describe, expect, test } from 'vitest';

import { Toaster } from './Toast';
import * as toastUtils from './utils';

function renderSampleApp() {
  return render(<Toaster />);
}

describe('Toast utils', () => {
  beforeEach(() => {
    renderSampleApp();
  });

  test('triggers default toast', async () => {
    toastUtils.toast('Default toast');
    const toast = await screen.findByRole('alert');
    within(toast).getByText('Default toast');
    within(toast).getByRole('button', { name: 'Close' });
  });

  test('triggers default toast with description', async () => {
    toastUtils.toast({
      title: 'Default toast',
      description: 'Description',
    });
    const toast = await screen.findByRole('alert');
    within(toast).getByText('Default toast');
    within(toast).getByText('Description');
    within(toast).getByRole('button', { name: 'Close' });
  });

  test('triggers default toast without close button', async () => {
    toastUtils.toast({
      title: 'Default toast',
      description: 'Description',
      isClosable: false,
    });

    const toast = await screen.findByRole('alert');
    within(toast).getByText('Default toast');
    within(toast).getByText('Description');
    expect(
      within(toast).queryByRole('button', { name: 'Close' }),
    ).not.toBeInTheDocument();
  });

  test('triggers default toast with action', async () => {
    const action = () => <button>Action</button>;
    toastUtils.toast({
      title: 'Default toast',
      description: 'Description',
      action,
    });

    const toast = await screen.findByRole('alert');
    within(toast).getByText('Default toast');
    within(toast).getByText('Description');
    within(toast).getByRole('button', { name: 'Action' });
    expect(
      within(toast).queryByRole('button', { name: 'Close' }),
    ).not.toBeInTheDocument();
  });

  test.each`
    utilVariantFunction         | variant
    ${toastUtils.toast.success} | ${'success'}
    ${toastUtils.toast.warning} | ${'warning'}
    ${toastUtils.toast.danger}  | ${'danger'}
  `('triggers $variant toast', async ({ utilVariantFunction }) => {
    utilVariantFunction({
      title: 'Title',
      description: 'Description',
    });
    const toast = await screen.findByRole('alert');
    within(toast).getByText('Title');
    within(toast).getByText('Description');
    within(toast).getByRole('button', { name: 'Close' });
    expect(toast).toMatchSnapshot();
  });
});
