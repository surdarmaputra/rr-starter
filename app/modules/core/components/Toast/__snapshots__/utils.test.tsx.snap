// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Toast utils > triggers 'danger' toast 1`] = `
<div
  class="relative flex w-full items-center gap-2 rounded-lg p-4 shadow-lg ring-1 md:w-80 bg-toast-danger-background ring-toast-danger-border shadow-danger-disabled"
  role="alert"
>
  <div
    class="h-6 w-6 shrink-0 self-start"
  >
    <svg
      class="lucide lucide-circle-alert h-6 w-6 text-white fill-toast-danger-title"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="12"
        cy="12"
        r="10"
      />
      <line
        x1="12"
        x2="12"
        y1="8"
        y2="12"
      />
      <line
        x1="12"
        x2="12.01"
        y1="16"
        y2="16"
      />
    </svg>
  </div>
  <div
    class="flex flex-1 items-center"
  >
    <div
      class="w-full"
    >
      <p
        class="pr-6 text-sm font-medium text-toast-danger-title"
      >
        Title
      </p>
      <p
        class="mt-1 pr-6 text-sm text-toast-danger-title opacity-60"
      >
        Description
      </p>
    </div>
  </div>
  <button
    class="absolute right-4 top-4 rounded-full bg-transparent p-1 text-toast-danger-title opacity-60 hover:bg-danger-disabled hover:text-toast-danger-title hover:opacity-100"
  >
    <span
      class="sr-only"
    >
      Close
    </span>
    <svg
      aria-hidden="true"
      class="lucide lucide-x h-4 w-4"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M18 6 6 18"
      />
      <path
        d="m6 6 12 12"
      />
    </svg>
  </button>
</div>
`;

exports[`Toast utils > triggers 'success' toast 1`] = `
<div
  class="relative flex w-full items-center gap-2 rounded-lg p-4 shadow-lg ring-1 md:w-80 bg-toast-success-background ring-toast-success-border shadow-success-disabled"
  role="alert"
>
  <div
    class="h-6 w-6 shrink-0 self-start"
  >
    <svg
      class="lucide lucide-circle-check h-6 w-6 text-white fill-toast-success-title"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="12"
        cy="12"
        r="10"
      />
      <path
        d="m9 12 2 2 4-4"
      />
    </svg>
  </div>
  <div
    class="flex flex-1 items-center"
  >
    <div
      class="w-full"
    >
      <p
        class="pr-6 text-sm font-medium text-toast-success-title"
      >
        Title
      </p>
      <p
        class="mt-1 pr-6 text-sm text-toast-success-title opacity-60"
      >
        Description
      </p>
    </div>
  </div>
  <button
    class="absolute right-4 top-4 rounded-full bg-transparent p-1 text-toast-success-title opacity-60 hover:bg-success-disabled hover:text-toast-success-title hover:opacity-100"
  >
    <span
      class="sr-only"
    >
      Close
    </span>
    <svg
      aria-hidden="true"
      class="lucide lucide-x h-4 w-4"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M18 6 6 18"
      />
      <path
        d="m6 6 12 12"
      />
    </svg>
  </button>
</div>
`;

exports[`Toast utils > triggers 'warning' toast 1`] = `
<div
  class="relative flex w-full items-center gap-2 rounded-lg p-4 shadow-lg ring-1 md:w-80 bg-toast-warning-background ring-toast-warning-border shadow-warning-disabled"
  role="alert"
>
  <div
    class="h-6 w-6 shrink-0 self-start"
  >
    <svg
      class="lucide lucide-triangle-alert h-6 w-6 text-white fill-toast-warning-title"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="m21.73 18-8-14a2 2 0 0 0-3.48 0l-8 14A2 2 0 0 0 4 21h16a2 2 0 0 0 1.73-3"
      />
      <path
        d="M12 9v4"
      />
      <path
        d="M12 17h.01"
      />
    </svg>
  </div>
  <div
    class="flex flex-1 items-center"
  >
    <div
      class="w-full"
    >
      <p
        class="pr-6 text-sm font-medium text-toast-warning-title"
      >
        Title
      </p>
      <p
        class="mt-1 pr-6 text-sm text-toast-warning-title opacity-70"
      >
        Description
      </p>
    </div>
  </div>
  <button
    class="absolute right-4 top-4 rounded-full bg-transparent p-1 text-toast-warning-title opacity-70 hover:bg-warning-disabled hover:text-toast-warning-title hover:opacity-100"
  >
    <span
      class="sr-only"
    >
      Close
    </span>
    <svg
      aria-hidden="true"
      class="lucide lucide-x h-4 w-4"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M18 6 6 18"
      />
      <path
        d="m6 6 12 12"
      />
    </svg>
  </button>
</div>
`;
